conans/client/graph/build_mode.py

```diff
@@ -44,9 +46,25 @@ def __init__(self, params, output):
 
             if self.never and (self.outdated or self.missing or self.patterns or self.cascade):
                 raise ConanException("--build=never not compatible with other options")
-        self._unused_patterns = list(self.patterns)
+        self._excluded_patterns = [it for it in list(self.patterns) if it.startswith("!")]
+        self._unused_patterns = [it for it in list(self.patterns)
+                                 if it not in self._excluded_patterns]
 
     def forced(self, conan_file, ref, with_deps_to_build=False):
+        def pattern_match(pattern):
+            return (fnmatch.fnmatchcase(ref.name, pattern) or
+                    fnmatch.fnmatchcase(repr(ref.copy_clear_rev()), pattern) or
+                    fnmatch.fnmatchcase(repr(ref), pattern))
+
+        for pattern in self.patterns:
```

Comment:
Shouldn't this use ``self._excluded_patterns``? Seems a bit confusing.

Reply:
It's a possibility indeed, but the idea here is comparing if a pattern matches with excluded pattern, if so, exclude it from exclude patterns list and return False to say: Don't build this package from sources. All remaining patterns in excluded list will be printed to say: These packages don't match to a pattern to excluded.